tidy3d/components/data_new.py

```diff
@@ -0,0 +1,137 @@
+
+    component: ComponentType = pd.Field(
+        None,
+        title="Original component",
+        description="Copy of the tidy3d component that this data is associated with.",
+    )
+
+    data: xr.DataArray = pd.Field(
+        None,
+        title="Data",
+        description="An xarray DataArray corresponding to the data stord by this element.",
+    )
+
+    data_dict: Dict[str, "Tidy3dData"] = pd.Field(
+        {},
+        title="Data dictionary",
+        description="Collection of other tidy3d data stored by this element",
+    )
+
+    @property
+    def coords(self) -> Dict[str, ArrayLike]:
+        """Labelled coordinates corresponding to the axes of ``self.values``."""
+        return self.data.coords
+
+    @property
+    def values(self) -> ArrayLike:
+        """The raw data associated with the data, represented as a multi-dimensional array."""
+        return self.data.values
+
+    @property
+    def dims(self) -> Tuple[str]:
+        """Tuple of the ordered coordinate labels corresponding to each axis of ``self.values.``"""
+        return self.data.dims
+
+    def __getitem__(self, name) -> "Tidy3dData":
+        """Return the ``Tidy3dData`` associated with the collection member using square brackets."""
+        item = self.data_dict.get(name)
+        if item is None:
+            raise Tidy3dKeyError(f'item "{item}" not found in data_dict.')
+        return item
+
+    @property
+    def json_string(self) -> str:
+        """A string representation of everything execpt for the ``.data`` and ``.data_dict``."""
+        return self.json(exclude={"data", "data_dict"})
+
+    """ Saving to an hdf5 file."""
+
+    def save_component(self, f_handle, hdf5_grp):
+        """Convert everything but the ``.data`` and ``.data_dict`` to a string and save to group."""
+        hdf5_grp.create_dataset("json", data=self.json_string)
+
+    def save_data(self, f_handle, hdf5_grp):
+        """Save the ``xarray.DataArray`` to an hdf5 group."""
+        hdf5_grp.create_dataset("values", data=self.values)
+        for dim in self.dims:
+            coord = self.coords[dim]
+            hdf5_grp.create_dataset(dim, data=coord)
+
+    def save_data_dict(self, f_handle, hdf5_grp):
+        """Save the ``data_dict`` to an hdf5 group recursively."""
+        for name, tidy3d_data in self.data_dict.items():
+            new_group = f_handle.create_group(name)
+            tidy3d_data.add_to_hdf5_group(f_handle, new_group)
+
+    def add_to_hdf5_group(self, f_handle, hdf5_grp) -> "hdf5_grp":
+        """Save the entire object to an hdf5 group."""
+        self.save_component(f_handle, hdf5_grp)
+        self.save_data(f_handle, hdf5_grp)
+        self.save_data_dict(f_handle, hdf5_grp)
+
+    def to_file(self, fname: str) -> None:
+        """Save the entire object to an hdf5 file."""
+        with h5py.File(fname, "a") as f_handle:
+            root_grp = f_handle.create_group("root")
+            self.add_to_hdf5_group(f_handle, root_grp)
```

Comment:
`f_handle` shouldn't be needed anywhere outside of this function.

Reply:
I think it's needed to create groups, but could be wrong.  So it would be needed to do the recursion in the data_dict part.